src/scalar_weighting.rs

```diff
@@ -0,0 +1,232 @@
+use std::borrow::Borrow;
+
+use tch::nn::{Init, Linear, Module, Path};
+use tch::{Kind, Reduction, Tensor};
+
+use crate::cow::CowTensor;
+use crate::traits::LayerOutput;
+
+/// Layer that performs a scalar weighting of layers.
+///
+/// Following Peters et al., 2018 and Kondratyuk & Straka, 2019, this
+/// layer applies scalar weighting:
+///
+/// *e = c ∑_i[ h_i · softmax(w)_i ]*
+#[derive(Debug)]
+pub struct ScalarWeight {
+    /// Layer dropout probability.
+    layer_dropout_prob: f64,
+
+    /// Layer-wise weights.
+    layer_weights: Tensor,
+
+    /// Scalar weight.
+    scale: Tensor,
+}
+
+impl ScalarWeight {
+    pub fn new<'a>(vs: impl Borrow<Path<'a>>, n_layers: i64, layer_dropout_prob: f64) -> Self {
+        assert!(
+            n_layers > 0,
+            "Number of layers ({}) should be larger than 0",
+            n_layers
+        );
+
+        assert!(
+            layer_dropout_prob >= 0.0 && layer_dropout_prob < 1.0,
+            "Layer dropout should be in [0,1), was: {}",
+            layer_dropout_prob
+        );
+
+        let vs = vs.borrow();
+
+        ScalarWeight {
+            layer_dropout_prob,
+            layer_weights: vs.var("layer_weights", &[n_layers], Init::KaimingUniform),
```

Comment:
Not sure whether random init is the way to go for the scalar weights. 

These all use zero init:

https://github.com/allenai/bilm-tf/blob/7cffee2b0986be51f5e2a747244836e1047657f4/bilm/elmo.py#L68
https://github.com/allenai/allennlp/blob/master/allennlp/modules/scalar_mix.py
https://github.com/Hyperparticle/udify/blob/master/udify/modules/scalar_mix.py

Reply:
I will make an issue for this, I guess the difference should be evaluated?

Reply:
I don't think it should be randomly initialized, the random init will give some layers more weight than others, by chance the low weight ones can be the ones that should have a high weight. With frozen encoders, it's probably not too bad, though.

Reply:
Also, not opposed to trying.

Reply:
Right. Zero-initialized is uniformly initialized. So I'll replace this.

Reply:
Fixed.